pkg/quota/admission/clusterresourcequota/accessor.go

```diff
@@ -0,0 +1,161 @@
+		// this should never happen
+		panic(err)
+	}
+
+	return &quotaAccessor{
+		quotaLister:     quotaLister,
+		namespaceLister: namespaceLister,
+		quotaClient:     quotaClient,
+		quotaMapper:     quotaMapper,
+		updatedQuotas:   updatedCache,
+	}
+}
+
+func (e *quotaAccessor) UpdateQuotaStatus(newQuota *kapi.ResourceQuota) error {
+	quota, err := e.quotaLister.Get(newQuota.Name)
+	if err != nil {
+		return err
+	}
+	quota = e.checkCache(quota)
+
+	// make a copy
+	obj, err := kapi.Scheme.Copy(quota)
+	if err != nil {
+		return err
+	}
+	quota = obj.(*quotaapi.ClusterResourceQuota)
+	usageDiff := utilquota.Subtract(newQuota.Status.Used, quota.Status.Total.Used)
+
+	// re-assign objectmeta
+	quota.ObjectMeta = newQuota.ObjectMeta
+	quota.Namespace = ""
+
+	oldNamespaceTotals, _ := quota.Status.Namespaces.Get(newQuota.Namespace)
+	namespaceTotalCopy, err := kapi.Scheme.DeepCopy(oldNamespaceTotals)
+	if err != nil {
+		return err
+	}
+	newNamespaceTotals := namespaceTotalCopy.(kapi.ResourceQuotaStatus)
+	newNamespaceTotals.Used = utilquota.Add(newNamespaceTotals.Used, usageDiff)
+	quota.Status.Namespaces.Insert(newQuota.Namespace, newNamespaceTotals)
+
+	quota.Status.Total.Used = utilquota.Add(quota.Status.Total.Used, usageDiff)
+
+	updatedQuota, err := e.quotaClient.ClusterResourceQuotas().Update(quota)
+	if err != nil {
+		return err
+	}
+
+	e.updatedQuotas.Add(quota.Name, updatedQuota)
+	return nil
+}
+
+var etcdVersioner = etcd.APIObjectVersioner{}
+
+// checkCache compares the passed quota against the value in the look-aside cache and returns the newer
+// if the cache is out of date, it deletes the stale entry.  This only works because of etcd resourceVersions
+// being monotonically increasing integers
+func (e *quotaAccessor) checkCache(quota *quotaapi.ClusterResourceQuota) *quotaapi.ClusterResourceQuota {
+	uncastCachedQuota, ok := e.updatedQuotas.Get(quota.Name)
+	if !ok {
+		return quota
+	}
+	cachedQuota := uncastCachedQuota.(*quotaapi.ClusterResourceQuota)
+
+	if etcdVersioner.CompareResourceVersion(quota, cachedQuota) >= 0 {
+		e.updatedQuotas.Remove(quota.Name)
+		return quota
+	}
+	return cachedQuota
+}
+
+func (e *quotaAccessor) GetQuotas(namespaceName string) ([]kapi.ResourceQuota, error) {
+	var quotaNames []string
+	// wait for a valid mapping cache.  The overall response can be delayed for up to 10 seconds.
+	err := utilwait.PollImmediate(100*time.Millisecond, 8*time.Second, func() (done bool, err error) {
+		var namespacelabels map[string]string
+		quotaNames, namespacelabels = e.quotaMapper.GetClusterQuotasFor(namespaceName)
+		namespace, err := e.namespaceLister.Get(namespaceName)
+		if err != nil {
+			return false, err
+		}
+		if kapi.Semantic.DeepEqual(namespacelabels, namespace.Labels) {
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	resourceQuotas := []kapi.ResourceQuota{}
+	for _, quotaName := range quotaNames {
+		quota, err := e.quotaLister.Get(quotaName)
```

Comment:
this is really confusing to read when types are hidden without remembering that quotaLister deals with ClusterResourceQuota and that quota in that case is of that type.  so anything we can do that is short convention to know that this quota is a crq and not rq would be very much appreciated.